### PR TITLE
fix: properly add CSS custom properties to inline style attributes

### PR DESCRIPTION
**Pull request by blittle**
## Details
Resolves https://github.com/salesforce/lwc/issues/2196

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅ 
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅ 

## GUS work item
N/A

### CODE REVIEW

**Comment by blittle** [comment]
If we consider the compiled output from LWC to be semantically versioned, then this _would_ have a breaking change.

**Comment by diervo** [comment]
My recollection is that we needed to do the IDL computed properties rather than the setAttribute for a reason, but I don't quite remember if it was perf or some off weird cases. @pmdartus will remember.

**Comment by blittle** [comment]
@diervo If that's the case, the best solution I think would be to add logic to the compiler and engine that tests if it is a custom property (`.startsWith('--')`) and avoid IDL and use `setProperty` if it's custom.

**Comment by diervo** [comment]
Im fine either way, let's see what Pierro says once he wakes up (unless someone else from the team remembers)

**Review by pmdartus — CHANGES_REQUESTED** [review]
@diervo I don't recall exactly the motivation behind using the style named properties instead of `setProperty`. I did some investigation related to `CSSStyleDeclaration.prototype.setProperty`:
- it is widely supported (including IE11). Attempting to set a CSS custom property using this method is a no-op on IE11.
- it is as performant as using a named properties setter.

Thanks for opening this PR @blittle, this is great. Could you please add a karma integration test to prevent future regression? The best place would be to add a new test case under `packages/integration-karma/test/template/attribute-style`

@Sumitjg Could you check why the CI is not picking up this PR originated for a fork?

**Comment by Sumitjg** [comment]
Sure @pmdartus .

**Comment by blittle** [comment]
@pmdartus updated

File: packages/integration-karma/test/template/attribute-style/index.spec.js
Comment (pmdartus) [inline packages/integration-karma/test/template/attribute-style/index.spec.js:14]: This will certainly fail on IE11.

**Review by pmdartus — APPROVED** [review]
LGTM once the CI is up again on forked PR. I would expect the test to fail on IE11 specifically. Don't worry I will make the change myself before merging.

Thanks again for the PR.

File: packages/integration-karma/test/template/attribute-style/index.spec.js
Comment (blittle) [reply packages/integration-karma/test/template/attribute-style/index.spec.js:14]: I'm curious to see your strategy to add tests that behave differently per browser.

**Comment by ravijayaramappa** [comment]
@blittle Found this article on circleci which talks about similar problem. It is covered in this article: https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-
Can you please login to circleci and go to your projects dashboard and follow salesforce/lwc repo? The page looks like this:
Screen Shot 2021-01-21 at 7 55 11 AM

**Comment by blittle** [comment]
@ravijayaramappa done, force pushed to retrigger a build. Is this always required? Does the CI not run properly for external devs that make PRs?

**Comment by blittle** [comment]
I think I ran the tests on an older version, looks like the current version has no `.setProperty` implementation for the vdom

**Comment by ravijayaramappa** [comment]
@blittle The test seem to fail in server side rendering mode. It appears that @lwc/engine-server does not support `.setProperty` api yet. You can look at the code [here](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine-server/src/renderer.ts#L266-L317). cc @pmdartus 


> Is this always required?
Based on the circleci documentation, it appears to be an one time operation. As long as you follow lwc repo, circleci will run your lwc PRs against our project.

> Does the CI not run properly for external devs that make PRs?
It does run. We recently updated our CircleCI config, that might also be requiring this additional step. We have tested our CI against forked repos. Our team members had not encountered this issue probably because all of us were already following the project. So thank you for discovering this issue. We will update our contributing documentation.

**Comment by blittle** [comment]
@ravijayaramappa thank you for the followup. I'll see if I can add an SSR impl for setProperty.

**Comment by pmdartus** [comment]
Now that we are moving toward using `CSSStyleDeclaration.prototype.setProperty` instead of using named properties setters we should be able to get rid of the `Proxy` in the SSR renderer.

Taking a step back, I also think that the `getStyleDeclaration` on the `Renderer` interface doesn't make much sense anymore since we know that the only operation done on the `CSSStyleDeclaration` is invoking `setProperty`. I would then recommend replacing the `getStyleDeclaration` method with `setCSSStyleProperty`:

```diff
export interface Renderer<N = HostNode, E = HostElement> {
    // ...
-   getStyleDeclaration(element: E): CSSStyleDeclaration;
+   setCSSStyleProperty(element: E, name: string, value: string): void;
}
```

Now speaking about performance, CSS properties are set in a loop. The proposed change also means that for each iteration now have to look up `Element.prototype.style`. The performance impact is marginal compared to the invocation cost of `setProperty` (less than 5% based on [this benchmark](https://esbench.com/bench/600a88cbd9c27600a7bfdb3f)). Considering that using the `style` attribute directly in the template is a corner case, the benefit of clear decoupling outweighs the perf impact.

File: packages/@lwc/engine-core/src/framework/modules/computed-style-attr.ts
Comment (blittle) [inline packages/@lwc/engine-core/src/framework/modules/computed-style-attr.ts:22]: Because `setCSSStyleProperty` always adds, when updating the whole style attribute, it needs to be cleared out first.

File: packages/integration-karma/test/template/attribute-style/index.spec.js
Comment (blittle) [inline packages/integration-karma/test/template/attribute-style/index.spec.js:11]: Just do a feature check to detect if the current browser is IE

**Comment by blittle** [comment]
I'm unsure why the integration tests are failing on circle ci. I don't think it is related to my changes.

**Review by pmdartus — APPROVED** [review]

**Comment by pmdartus** [comment]
@Sumitjg Integration tests are still not passing from this forked branch. The build fails because the SauceLabs keys are not properly injected. I am also suspecting a similar issue with Best not running. Could you investigate this?

**Comment by ravijayaramappa** [comment]
@blittle @pmdartus this PR is good to go from CI perspective :shipit:

**Review by ravijayaramappa — APPROVED** [review]
One minor question about type script, otherwise LGTM

File: packages/@lwc/engine-dom/src/renderer.ts
Comment (blittle) [reply packages/@lwc/engine-dom/src/renderer.ts:190]: typo, will fix

**Pull request merged**
